src/persist-types/src/dyn_struct.rs

```diff
@@ -0,0 +1,512 @@
+
+//! Type erased columnar structs.
+
+use std::collections::BTreeMap;
+use std::fmt::Debug;
+use std::sync::Arc;
+
+use arrow2::array::{Array, NullArray, StructArray};
+use arrow2::bitmap::MutableBitmap;
+use arrow2::datatypes::{DataType as ArrowLogicalType, Field};
+use arrow2::io::parquet::write::Encoding;
+
+use crate::columnar::sealed::{ColumnMut, ColumnRef};
+use crate::columnar::{ColumnGet, ColumnPush, Data, DataType};
+use crate::dyn_col::{DynColumnMut, DynColumnRef};
+use crate::stats::{DynStats, OptionStats, StatsFn, StructStats};
+
+/// The schema of a "dynamic" columnar struct.
+///
+/// Here, _dynamic_ is used to mean that, unlike most [Data] impls, the columnar
+/// schema of this type is not inferrable purely from the type. Instead, it
+/// needs supplementary [ColumnCfg].
+///
+/// This exists because, among other reasons, mz's `Row` type does not declare
+/// the schema at the type level (as opposed to something like `Row(u64, i32)`).
+#[derive(Debug, Clone)]
+#[cfg_attr(debug_assertions, derive(PartialEq))]
+pub struct DynStructCfg {
+    pub(crate) schema: Arc<Vec<(String, DataType, StatsFn)>>,
+}
+
+impl From<Vec<(String, DataType, StatsFn)>> for DynStructCfg {
+    fn from(value: Vec<(String, DataType, StatsFn)>) -> Self {
+        DynStructCfg {
+            schema: Arc::new(value),
+        }
+    }
+}
+
+/// A "dynamic" columnar struct.
+///
+/// See [DynStructCfg].
+#[derive(Debug)]
+pub struct DynStruct;
+
+/// The [Data::Ref] type for [DynStruct].
+#[derive(Debug)]
+pub enum DynStructRef<'a> {
+    /// A sentinel to push the default value of each field in the struct.
+    Default,
+    /// A specific offset in a struct column.
+    Idx {
+        /// The offset.
+        idx: usize,
+        /// The fields of the struct column being referenced.
+        cols: &'a [DynColumnRef],
+    },
+}
+
+impl Default for DynStructRef<'_> {
+    fn default() -> Self {
+        DynStructRef::Default
+    }
+}
+
+/// A [ColumnRef] and [ColumnGet] impl for [DynStruct].
+#[derive(Debug)]
+pub struct DynStructCol {
+    len: usize,
+    cfg: DynStructCfg,
+    pub(crate) validity: Option<<bool as Data>::Col>,
+    pub(crate) cols: Vec<DynColumnRef>,
+}
+
+impl ColumnRef<DynStructCfg> for DynStructCol {
+    fn cfg(&self) -> &DynStructCfg {
+        &self.cfg
+    }
+    fn len(&self) -> usize {
+        self.len
+    }
+    fn to_arrow(&self) -> (Encoding, Box<dyn Array>) {
+        let array: Box<dyn Array> = match self.to_arrow_struct() {
+            Some((array, _col_encodings)) => Box::new(array),
+            None => Box::new(NullArray::new_empty(ArrowLogicalType::Null)),
+        };
+        (Encoding::Plain, array)
+    }
+    fn from_arrow(cfg: &DynStructCfg, array: &Box<dyn Array>) -> Result<Self, String> {
+        Self::from_arrow(cfg, array)
+    }
+}
+
+impl ColumnGet<DynStruct> for DynStructCol {
+    fn get<'a>(&'a self, idx: usize) -> DynStructRef<'a> {
+        assert!(self.validity.is_none());
```

Comment:
meant to say this when I first sent the PR out, all these ColumnGet and ColumnPush impls for DynStruct are unused, becuase Row and SourceData unpack the columnar struct into components (fields + validity) for performance via `DynStructCol::as_opt_ref` and `DynStructMut::as_opt_mut`. I implemented them because I wanted to prove to myself that we could, but AFAICT even a codepath for Datam::Record wouldn't use them, so they're dead code. should I replace them with something like `unreachable!` to make things more clear?